Extend pasta prepare_evaluation to handle review analysis

anmolsingh20 opened the pull request:
Hi Ralf,

Here is a concise script form of the first basic filtering and analysis of patch review. I could rebase it on the newest next once the review is through.

Many thanks,
Anmol

Commit:
Extend pasta prepare_evaluation to handle review analysis

It is essential to break the processing of review data into several steps
to avoid memory issues. The review analysis has five stages in the order:
prepare, preprocess, merge, filter, analyze. Further analyses on
review data can be integrated in a similar modular way.


Signed-off-by: Anmol Singh <anmol.singh@bmw.de>

rralf commented:
Hi Anmol,

can you give me an example how to start or how to run your scripts? Will leave some further comments in the commit.

Thanks!

rralf commented on analyses/response_analysis.py:
```diff
@@ -0,0 +1,151 @@
+
```
missing copyright header

rralf commented on analyses/response_analysis.py:
```diff
@@ -0,0 +1,151 @@
+import dask.dataframe as dd
+import matplotlib.pyplot as plt
+import pandas as pd
+import seaborn as sns
```
If you introduce new dependencies, please maintain the docker scripts. For CI purposes, PaStA needs to be able to run without any issues with the docker image.

anmolsingh20 replied:
Sure, I will do that.

rralf commented on analyses/response_analysis.py:
```diff
@@ -0,0 +1,151 @@
+    plt.savefig('patch_with_upstream_dist_response_counts.pdf')
+    plt.close()
+
+    log.info("Saved plots, done!")
```
Note: I didn't review the rest, I'd like to run it first.

anmolsingh20 replied:
Ok

rralf commented on bin/pasta_prepare_evaluation.py:
```diff
@@ -13,11 +13,23 @@
+import flat_table
+import numpy as np
 import os
+import pandas as pd
```
Same here, please update the docker files and test if the image is still able to run pasta.

anmolsingh20 replied:
Sure

rralf commented on bin/pasta_prepare_evaluation.py:
```diff
@@ -26,8 +38,7 @@
 
-from pypasta.LinuxMaintainers import LinuxMaintainers
-from pypasta.LinuxMailCharacteristics import load_linux_mail_characteristics
+sys.path.append(os.path.join(os.path.abspath(os.pardir), 'analyses'))
```
I wonder if there's a better way to do the imports than the current sys.path.append...-hack that we do in PaStA... I need to check that.

anmolsingh20 replied:
Yes, I thought about it too. Unfortunately this is the cleanest I could make it. Reason: We have under analyses, mixture of R and python, due to which it is not a package/module. Imports are unclean when we have a mix.

rralf commented on bin/pasta_prepare_evaluation.py:
```diff
@@ -343,33 +353,229 @@ def _load_responses_dict(msg_id, response_list):
+        else:
+            return None
+
+    dd1 = dd.read_csv(config.f_denorm_responses, blocksize=1e9, dtype={"idx ": "int32", "patch_id ": "category",
```
So if we introduce dask + pandas as a dependency, we should use it elsewhere as well, where we can benefit. E.g., the startup of PaStA is pretty slow when loading the indexes. I'll check if we can use pandas for reading the csvs - so if we introduce pands, it'd be great if there's a chance to boost or improve other things with it

rralf replied:
BTW, what influence does the blocksize have?

anmolsingh20 replied:
> BTW, what influence does the blocksize have?

This is to specify how you chunk your csv

anmolsingh20 replied:
> So if we introduce dask + pandas as a dependency, we should use it elsewhere as well, where we can benefit. E.g., the startup of PaStA is pretty slow when loading the indexes. I'll check if we can use pandas for reading the csvs - so if we introduce pands, it'd be great if there's a chance to boost or improve other things with it

It is a good idea to test this.

rralf replied:
> > BTW, what influence does the blocksize have?
> 
> This is to specify how you chunk your csv

chunk for what? As a test, I tried to, for example, parse indexes with panda (it's indeed faster), but I didn't see any influence of the blocksize, as there's no parallelisation or sth. similar (afaict).

anmolsingh20 replied:
> > > BTW, what influence does the blocksize have?
> > 
> > 
> > This is to specify how you chunk your csv
> 
> chunk for what? As a test, I tried to, for example, parse indexes with panda (it's indeed faster), but I didn't see any influence of the blocksize, as there's no parallelisation or sth. similar (afaict).

A dask dataframe is a made up of pandas dataframes (indexed and split). Blocksize affects these partitions when csv is ingested as a dask dataframe. It has to be chosen to balance good resource usage - cpus, memory, and scheduling overhead. I played around with it when my data was gigantic. I cannot however justify this value, may be it is an artifact of some of the trials I made on some dataframe along the way. May be the defaults are good enough. We have to check this.

rralf replied:
Ah, ok. Thanks for the explanation! Had similar issues with the parallelisation/Pool library of python: you can specify a worksize and a chunksize. The weight of those values have massive influence on memory consumption, however, one should have a look into a magic crystal ball to find sane values :-)

rralf commented on bin/pasta_prepare_evaluation.py:
```diff
@@ -343,33 +353,229 @@ def _load_responses_dict(msg_id, response_list):
+def _is_response_from_bot(message_id):
+    message = _repo.mbox.get_messages(message_id)[0]
+    email = email_get_from(message)[1].lower()
+    bots = ['tip-bot2@linutronix.de', 'tipbot@zytor.com', 'noreply@ciplatform.org', 'syzbot',
```
Why didn't you use / improve the stuff that we have in LinuxMailcharacteristics.py?

anmolsingh20 replied:
I had to extract this part out, as the method _is_from_bot in LinuxMailcharacteristics is tightly coupled with the class. This prevents me from using it when a message is not a patch (e.g. a response message) and things like is_patch, or _is_next are irrelevant for a response. This is not how it should finally look, and I look forward to your thoughts how to combine this best. I have to also admit, I do not fully understand all functionality of LinuxMailcharacteristics regarding the domain.

Looking forward to your remarks here.

rralf replied:
Yes, that's correct. At the moment it's tightly coupled. @ditzlike had a similar issue -- she wanted to use MailCharacteristics to assign a mail to the latest kernel version that released at that point in time. In this case, we had tight coupling, but in fact there's no need to instantiate a MailCharacteristics instance. In her case, it was possible to factor out _patch_get_version(): ref. b429ec028860fbe095cbc9f8d741abe264f58acf .

Maybe it's (somehow) possible to convert _is_from_bot() to a staticmethod.

anmolsingh20 replied:
I used LinuxMailCharacteristics and improved the _is_from_bot method in two new patches. Rebased to next as well to include your changes to the class.

rralf commented on bin/pasta_prepare_evaluation.py:
```diff
@@ -343,33 +353,229 @@ def _load_responses_dict(msg_id, response_list):
-                        dest='mode',
+    parser.add_argument('--offlist',
+                        action='store_true',
+                        default=False,
```
Hmm, I guess we should use the same scheme that we used in Basak's script:
--mode offlist, --mode ignored, … or something like that.

anmolsingh20 replied:
Ok, I will have a look. I made the change because I wanted to mix this with heterogeneous choices per switch. It should be possible though.

rralf commented on bin/pasta_prepare_evaluation.py:
```diff
@@ -343,33 +353,229 @@ def _load_responses_dict(msg_id, response_list):
     analysis_option = parser.parse_args(argv)
 
-    if not analysis_option.mode:
+    if len(argv) == 0:
```
... Then this is implicitly solved, as the parser will fail in case we didn't provide a choice.

rralf reviewed (changes requested):
Please have a look at the comments. Thanks!

rralf committed:
pypasta: Mbox: improve handling of raw mboxes

We can treat all kinds of mboxes the same, we just need some adjustments for
raw mboxes.

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

rralf committed:
pypasta: Mbox: simplify infrastructure of public inboxes

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

rralf committed:
pypasta: Mbox: simplify patchwork handling

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

rralf committed:
pypasta: Mbox: consolidate stuff

Now everything is in place to get a common handling.

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

rralf committed:
pypasta: Mbox: simplify add_mail_to_list

Only one user is left, inline and simplify it.

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

rralf committed:
pypasta: Util: introduce path_convert_relative

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

anmolsingh20 commented:
> Hi Anmol,
> 
> can you give me an example how to start or how to run your scripts? Will leave some further comments in the commit.
> 
> Thanks!

Sure,
./pasta prepare_evaluation --review (prepare, preprocess, merge, filter, analyse), in that order. This ensures you have the required intermediate results, and do not hold more data in memory at once.

rralf committed:
pypasta: Mbox: reduce usages of os.path by direct import

shorter lines are better lines.

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

rralf commented:
> > Hi Anmol,
> > can you give me an example how to start or how to run your scripts? Will leave some further comments in the commit.
> > Thanks!
> 
> Sure,
> ./pasta prepare_evaluation --review (prepare, preprocess, merge, filter, analyse), in that order. This ensures you have the required intermediate results, and do not hold more data in memory at once.

Thanks, will test it soon!

rralf committed:
resources: forward resources

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

rralf committed:
pypasta: LinuxMailCharacteristics: remove unused argument

no functional change.

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

rralf committed:
pypasta: LinuxMailCharacteristics: allow instantiation w/o clustering and maintainers

Signed-off-by: Ralf Ramsauer <ralf.ramsauer@oth-regensburg.de>

Commit:
pypasta: LinuxMailCharacteristics: simplify _is_from_bot

no functional change; improve formatting

Signed-off-by: Anmol Singh <anmol.singh@bmw.de>

Commit:
Use LinuxMailCharacteristics bot logic in review analysis

LinuxMailCharacteristics can now be used for messages which are not
patches, hence remove duplicate logic for responses/patches and use the
common _is_from_bot function.


Signed-off-by: Anmol Singh <anmol.singh@bmw.de>

Commit:
Add new dependencies to docker script for patch review evaluation

flat-table (for normalizing dicts in pandas with NaN support),
pandas for python data analytics,
dask for scaling pandas for huge datasets, matplotlib and seaborn
for plotting in python.


Signed-off-by: Anmol Singh <anmol.singh@bmw.de>

rralf commented on pypasta/LinuxMailCharacteristics.py:
```diff
@@ -92,15 +91,15 @@
                  'netfilter-devel@vger.kernel.org', 'platform-driver-x86@vger.kernel.org',
                  'reiserfs-devel@vger.kernel.org', 'selinux@vger.kernel.org', 'sparclinux@vger.kernel.org',
-                 'stable@vger.kernel.org','target-devel@vger.kernel.org', 'util-linux@vger.kernel.org',
+                 'stable@vger.kernel.org', 'target-devel@vger.kernel.org', 'util-linux@vger.kernel.org',
```
Yep, will merge this. By the way: This list ist derived from the public inboxes. Note to self: We added new mailing lists, so this list might be incomplete. Will have to check that.

rralf commented on pypasta/LinuxMailCharacteristics.py:
```diff
@@ -152,12 +151,12 @@ def __init__(self, c):
         # Metric: At least one correct list + at least one correct maintainer
         if (not c.mtrs_has_lists or c.mtrs_has_one_correct_list) and \
-           (not c.mtrs_has_maintainers or c.mtrs_has_one_correct_maintainer):
+                (not c.mtrs_has_maintainers or c.mtrs_has_one_correct_maintainer):
```
it this the "correct" indentation?

anmolsingh20 replied:
I did reformat code in pycharm, which has default settings for python.

rralf commented on pypasta/LinuxMailCharacteristics.py:
```diff
@@ -221,48 +220,35 @@ class LinuxMailCharacteristics:
 
+        if email in bots:
+            return True
+        elif email in potential_bots and \
```
No need for the elif, as we would have returned before. Will change this during merge.

rralf commented on pypasta/LinuxMailCharacteristics.py:
```diff
@@ -221,48 +220,35 @@ class LinuxMailCharacteristics:
-            return True
-
+        bots = ['tip-bot2@linutronix.de', 'tipbot@zytor.com', 'noreply@ciplatform.org', 'syzbot',
+                'syzkaller.appspotmail.com']
```
Ah, thanks for picking this up! We have another undetected bot left: The Hulk Robot <hulkci@huawei.com>

rralf replied:
... And I'll move the constants to the top of the class.

rralf commented on pypasta/LinuxMailCharacteristics.py:
```diff
@@ -221,48 +220,35 @@ class LinuxMailCharacteristics:
-
-        return False
+        else:
+            return False
```
Thanks. I'll now carefully test this change with a bunch of emails.

rralf commented on pypasta/LinuxMailCharacteristics.py:
```diff
@@ -221,48 +220,35 @@ class LinuxMailCharacteristics:
-           email_get_header_normalised(message, 'x-patchwork-hint') == 'ignore':
-            return True
-
+        bots = ['tip-bot2@linutronix.de', 'tipbot@zytor.com', 'noreply@ciplatform.org', 'syzbot',
```
Will change this to a set.

anmolsingh20 replied:
ok

rralf commented on docker/pasta-skeleton.dockerfile:
```diff
@@ -42,7 +42,10 @@ RUN apt install -y --no-install-recommends \
+	flask-nav anytree \
+	flat-table \
+	pandas "dask[complete]" \
+	matplotlib seaborn
```
Don't we get those libraries via Ubuntu's official repository? I would prefer that over pip for stability reasons.

rralf commented on pypasta/LinuxMailCharacteristics.py:
```diff
@@ -221,48 +220,35 @@ class LinuxMailCharacteristics:
+        uagent = email_get_header_normalised(message, 'user-agent')
+        xmailer = email_get_header_normalised(message, 'x-mailer')
 
+        if email in bots:
```
This check is incorrect for 'syzbot', as it's not a valid email address. However, we to classify all mails with 'syzbot' in their mail adress as bots. Will fix this.

anmolsingh20 replied:
Ok, thanks

rralf commented:
I manually picked the "simplify _is_from_bot" patch, and added some changes and fixed the szybot-bug. Rebased your series and pushed it to wip-aggregation-analyses. Still need to review the rest...
